### PR TITLE
IGNITE-16331 Adds topology validator implementation.

### PR DESCRIPTION
**Pull request by ololo3000**

### CODE REVIEW

File: modules/topology-validator-ext/src/test/java/com/sbt/ignite/cache/IgniteCacheTopologyValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+    /** */
+    public static final int CACHE_CNT = 2;
+
+    /** */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return getConfiguration(igniteInstanceName, true);
+    }
+
+    /** */
+    private IgniteConfiguration getConfiguration(
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/test/java/com/sbt/ignite/cache/IgniteCacheTopologyValidatorTest.java:91]: Done.

File: modules/topology-validator-ext/pom.xml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>${ignite.groupId}</groupId>
+            <artifactId>ignite-spring</artifactId>
+            <version>${ignite.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+</project>
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/pom.xml]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    private void onGlobalClusterStateChanged(ClusterState clusterState) {
+        state = clusterState == ACTIVE ? State.VALID : State.CLUSTER_WRITE_BLOCKED;
+    }
+
+    /**
+     * Current implementation of segmentation detection compares node of each topology with configured baseline nodes.
+     * If baseline auto adjustment is configured with zero timeout - baseline is updated on each topology change
+     * and the comparison described above makes no sense.
+     */
+    private boolean checkBaselineAutoAdjustConfiguration(boolean enabled, long timeout) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        }
+
+        /** @return String representation of the specified cluster node collection. */
+        private String formatTopologyNodes(Collection<ClusterNode> nodes) {
+            return nodes.stream().map(n -> n.id().toString()).collect(Collectors.joining(", "));
+        }
+    }
+
+    /** Represents possible states of the current segment. */
+    private enum State {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+     * Current implementation of segmentation detection compares node of each topology with configured baseline nodes.
+     * If baseline auto adjustment is configured with zero timeout - baseline is updated on each topology change
+     * and the comparison described above makes no sense.
+     */
+    private boolean checkBaselineAutoAdjustConfiguration(boolean enabled, long timeout) {
+        if (isDisabled())
+            return false;
+
+        if (enabled && timeout == 0L) {
+            U.warn(log, "Topology Validator is currently skipping validation of topology changes because" +
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+                ctx.state().isBaselineAutoAdjustEnabled(),
+                ctx.state().baselineAutoAdjustTimeout()
+            );
+
+            if (!skipTopChangeCheck && state == State.VALID && evt.type() == EVT_NODE_FAILED) {
+                List<? extends BaselineNode> baselineNodes = discoCache.baselineNodes();
+
+                // Actually Ignite considers segmentation as the sequential node failures. So we detect segmentation
+                // even if the single node fails and less than half of baseline nodes are alive.
+                if (baselineNodes != null && aliveBaselineNodes(baselineNodes) < baselineNodes.size() / 2 + 1) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+            State locStateCopy = state;
+
+            lastCheckedTopVer = evt.topologyVersion();
+
+            boolean skipTopChangeCheck = !checkBaselineAutoAdjustConfiguration(
+                ctx.state().isBaselineAutoAdjustEnabled(),
+                ctx.state().baselineAutoAdjustTimeout()
+            );
+
+            if (!skipTopChangeCheck && state == State.VALID && evt.type() == EVT_NODE_FAILED) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: No, we should not. EVT_NODE_LEFT is raised on graceful shutdown and has nothing to do with segmentation.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        if (data == null) {
+            String msg = "The Topology Validator plugin is not configured for the server node that is" +
+                " trying to join the cluster. Since the Topology Validator is only applicable if all server nodes" +
+                " in the cluster have one, node join request will be rejected [rejectedNodeId=" + node.id() + ']';
+
+            throw new PluginValidationException(msg, msg, node.id());
+        }
+
+        // If the new node is joining but some node failed/left events has not been handled by
+        // {@linkTopologyChangedEventListener} yet, we cannot guarantee that the {@link state} on the joining node will
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+            for (BaselineNode node : baselineNodes) {
+                if (!(node instanceof DetachedClusterNode))
+                    ++res;
+            }
+
+            return res;
+        }
+
+        /** @return String representation of the specified cluster node collection. */
+        private String formatTopologyNodes(Collection<ClusterNode> nodes) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

File: modules/topology-validator-ext/src/test/java/com/sbt/ignite/cache/IgniteCacheTopologyValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+
+        CountDownLatch discoveryWorkerBlockedLatch = new CountDownLatch(1);
+
+        try {
+            srv.events().localListen(evt -> {
+                try {
+                    discoveryWorkerBlockedLatch.await();
+                }
+                catch (InterruptedException e) {
+                    U.error(log, e);
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/test/java/com/sbt/ignite/cache/IgniteCacheTopologyValidatorTest.java:431]: If this exception occurs, assertion below will fail the test. I don't think it's worth it to fail the test immediately after this exception, given that the listener code is running from an Ignite worker thread.

File: modules/topology-validator-ext/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/pom.xml]: Done.

File: modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+import static org.apache.ignite.cluster.ClusterState.ACTIVE_READ_ONLY;
+import static org.apache.ignite.events.EventType.EVT_NODE_FAILED;
+import static org.apache.ignite.events.EventType.EVT_NODE_JOINED;
+import static org.apache.ignite.events.EventType.EVT_NODE_LEFT;
+import static org.apache.ignite.internal.cluster.DistributedConfigurationUtils.setDefaultValue;
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.PUBLIC_POOL;
+
+/** */
+public class CacheTopologyValidatorPluginProvider implements PluginProvider<PluginConfiguration>, TopologyValidator {
+    /** */
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/topology-validator-ext/src/main/java/com/sbt/ignite/plugin/cache/CacheTopologyValidatorPluginProvider.java]: Done.

**Review by Mmuzaf — APPROVED** [review]

**Review by Mmuzaf — APPROVED** [review]

**Pull request merged**